File: JavaScript/Semana 4/Quick.js
function swap(a, i, j) {
    let temp = a[i];
    a[i] = a[j];
    a[j] = temp;
}

function partition(a, l, h){
    let pivot = a[h];
    let i = l - 1;

    for (let j = l; j <= h - 1; j++) {
        if (a[j] < pivot) {
            i++;
            swap(a, i, j);
        }
    }
    swap(a, i + 1, h);
    return (i + 1);
}

function quickSort(a, l, h) {
    if (l < h) {
        let p = partition(a, l, h);
        quickSort(a, l, p - 1);
        quickSort(a, p + 1, h);
    }
}

function printArr(a) {
    for (let i = 0; i < a.length; i++) {
        process.stdout.write(a[i] + " ");
    }
}

let a = [65, 26, 13, 23, 12];

console.log("Arreglo antes de ser ordenado:");
printArr(a);

quickSort(a, 0, a.length - 1);

console.log("\nArreglo despues de ser ordenado:");
printArr(a);

File: C++/Semana 4/Quick.cpp
#include <iostream>

void swap(int a[], int i, int j) {
    int temp = a[j];
    a[j] = a[i];
    a[i] = temp;
}

int partition(int a[], int l, int h) {
    int pivot = a[h];
    int i = (l - 1);

    for (int j = l; j <= h - 1; j++) {
        if (a[j] < pivot) {
            i++;
            swap(a, i, j);
        }
    }
    swap(a, i + 1, h);
    return (i + 1);
}

void quickSort(int a[], int l, int h) {
    if (l < h) {
        int pi = partition(a, l, h);
        quickSort(a, l, pi - 1);
        quickSort(a, pi + 1, h);
    }
}


void printArr(int a[], int size) {
    for (int i = 0; i < size; i++)
        std::cout << a[i] << " ";
    std::cout << std::endl;
}


int main() {
    int a[] = { 10, 7, 8, 9, 1, 5 };
    std::cout << "arreglo antes de ordenar: ";
    printArr(a, 6);

    quickSort(a, 0, 5);
    std::cout << "arreglo despues de ordenar: ";
    printArr(a, 6);

    return 0;
}